packages/doenetml-worker-rust/lib-doenetml-core/src/core/components/component_builder.rs

```diff
@@ -0,0 +1,164 @@
+            .nodes
+            .iter()
+            .enumerate()
+            .map(|(idx, node)| match node {
+                NormalizedNode::Element(elm) => {
+                    let mut component = ComponentEnum::from_str(&elm.name).unwrap_or_else(|_| {
+                        // If we didn't find a match, then create a component of type external
+                        ComponentEnum::_External(_External {
+                            name: elm.name.clone(),
+                            ..Default::default()
+                        })
+                    });
+                    if let Some(Source::Macro(_)) = &elm.extending {
+                        // Some components specify that when they are referenced with the `$foo` syntax,
+                        // a different component should be created in their place. E.g., `<textInput name="i" />$i`
+                        // should become `<textInput name="i" /><text extend="$i` />`.
+                        // This information is stored in `ref_transmutes_to()`.
+                        if let Some(name) = component.ref_transmutes_to() {
+                            // It is forbidden to expand to an invalid component type, so we do not
+                            // have a fallback to `_External` here.
+                            component = ComponentEnum::from_str(name).unwrap();
+                        }
+                    }
+                    // Some references
+                    component.ref_transmutes_to();
+                    // XXX: we temporarily fill each required attribute with an empty vector.
+                    // This will be removed when typed attributes are integrated.
+                    let attributes: HashMap<&'static str, _> = HashMap::from_iter(
+                        component
+                            .get_attribute_names()
+                            .iter()
+                            .map(|&name| (name, Vec::new())),
+                    );
+
+                    let extending: Result<Option<ExtendSource>, anyhow::Error> =
+                        match &elm.extending {
+                            Some(ref_resolution) => {
+                                let ref_resolution = ref_resolution.get_resolution();
+                                // If there is no remaining path, we are extending a component directly. Otherwise,
+                                // we look up the state variable on that component and extend it.
+                                if let Some(unresolved_path) = &ref_resolution.unresolved_path {
+                                    if unresolved_path.len() != 1 {
+                                        Err(anyhow!("Handle nested state variables"))
+                                    } else {
+                                        // We only know how to handle single state variable access at the moment.
+                                        if let Some(state_var) = component
```

Comment:
Use `.get_public_state_variable_index_from_name_case_insensitive()` as we can match only public state variables and paths are matched in a case-insensitive manner.